LaraApi/src-lara/lara/pass/Pass.js

```diff
@@ -43,18 +47,83 @@ class Pass {
   }
 
   /**
-   * @abstract
+   * Apply tranformation to
+   * @abstract Contains default implementation only if matchJoinPoint and transformJoinpoint are implemented.
    */
   _apply_impl($jp) {
+    if (
+      this.matchJoinpoint === Pass.prototype.matchJoinpoint ||
+      this.transformJoinpoint === Pass.prototype.transformJoinpoint
+    ) {
+      throw new AbstractClassError({
+        kind: "abstractMethod",
+        baseClass: Pass,
+        derivedClass: this.constructor,
+        method: this._apply_impl,
+      });
+    }
+
+    const matchingJps = [$jp, ...$jp.descendants].filter(($jp) =>
+      this.matchJoinpoint($jp)
+    );
+
+    let insertedLiteralCode = false;
+    let casesApplied = 0;
+    let transformationErrors = [];
+
+    for (const $jp of matchingJps) {
+      try {
+        const result = this.transformJoinpoint($jp);
+        casesApplied++;
+        insertedLiteralCode = insertedLiteralCode || result.insertedLiteralCode;
+      } catch (e) {
+        if (e instanceof PassTransformationError) {
+          transformationErrors.push(e);
+        } else throw e;
+      }
+    }
+
+    return AgregatePassResult({
+      pass: this.constructor,
+      casesFound: matchingJps.length,
+      casesApplied,
+      transformationErrors,
+      insertedLiteralCode,
+    });
+  }
+
+  _new_default_result() {
+    return new PassResult(this.name);
+  }
+
+  /**
+   * Predicate that informs the pass whether a certain joinpoint should be transformed.
+   * @abstract
+   * @param {joinpoint} $jp Join point to match
+   * @returns {boolean}
+   */
+  matchJoinpoint($jp) {
     throw new AbstractClassError({
       kind: "abstractMethod",
       baseClass: Pass,
       derivedClass: this.constructor,
-      method: this._apply_impl,
+      method: this.matchJoinpoint,
     });
   }
 
-  _new_default_result() {
-    return new PassResult(this.name);
+  /**
+   * Transformation to be applied to matching joinpoints
+   * @abstract
+   * @param {joinpoint} $jp Join point to transform
+   * @throws {PassTransformationError} If the transformation fails
+   * @return {PassTransformationResult} The result of the transformation
+   */
+  transformJoinpoint($jp) {
+    throw new AbstractClassError({
```

Comment:
Could this constructor be replaced by a static function in AbstractClassError that received a JS function of a class? (AbstractClassError.newMethodError(this.transformJoinpoint) - name is just an example )

 It seems the information could be inferred from there, but I'm not sure.